### PR TITLE
feat: Update all plugins result

### LINKED ISSUE #980: [Feature Request] - ability to bulk update plugins
It would be good to have the ability to tell Flow to update all outdated plugins with one command.
Perhaps a new command like `pm update all`

### PR DESCRIPTION
**Pull request by flooxo**
As described in #980, having the ability to run updates for all plugins would be nice.
Therefore when running `pm update` the top result is now the option to update all plugins that have an update available.

ToDo:
- [x] Adjust search result
- [x] Only show option if more than one update is available
- [x] Implement logic

Close #980

### CODE REVIEW

File: Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs
Comment (flooxo) [inline Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs:211]: Maybe someone has an idea how to make this prettier. It works but I guess setting the top result in that way is not the most elegant solution :)

File: Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs
Comment (flooxo) [inline Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs]: Maybe someone has an idea how to make it more elegant. It works to set the top result this way, but I guess it's not the most elegant way

**Comment by flooxo** [comment]
Can anyone give me a hint why the tests are failing :D
Is it because I have a branch and the changes are not made on dev?

**Comment by VictoriousRaptor** [comment]
> Can anyone give me a hint why the tests are failing :D Is it because I have a branch and the changes are not made on dev?

Spell check is broken when checking cross-repo PRs now. It's not mandatory so take it easy.

File: Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs:266]: Please remove :)

**Comment by jjw24** [comment]
There is work to allow users to install/update/uninstall multiple plugins  https://github.com/Flow-Launcher/Flow.Launcher/pull/2369, just a heads up to keep in mind as it is close to finishing soon and it's making changes in the same area as this PR. I will be merging that one first before this PR, so watch out for code conflicts.

**Comment by flooxo** [comment]
Thanks for the hint

**Comment by VictoriousRaptor** [comment]
I did some major refactor in #2369 so please check that PR.

**Comment by flooxo** [comment]
Oke, thanks, then I'll wait until #2369 is merged and then I'll continue to keep it simple

**Comment by VictoriousRaptor** [comment]
> Oke, thanks, then I'll wait until #2369 is merged and then I'll continue to keep it simple Are translations added automatically by Crowdin, or is there a command for this?

Translation are done by community on crowdin. After merging to dev crowdin will add xaml keys for other languages so the community can translate.

File: Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs
Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs:266]: I think he means to remove the `TODO` comment once you finish it.

Comment (jjw24) [reply Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs:308]: Yeah, please remove the todo comment or do the todo please.

File: Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs
Comment (flooxo) [inline Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs:305]: Currently `pm update all plugins` is displayed if more than 2 updates are available. How can I change this so that only `pm update` but the translation is still displayed as the title of the result?

Comment (flooxo) [reply Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs:305]: Nope, the original idea was just to call `pm update` and show the option to run all updates to avoid introducing a new argument/command

Comment (jjw24) [reply Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs:305]: >the translation is still displayed as the title of the result?

Sorry I am also not quite understanding this line. Are you after a translated result title?

Comment (flooxo) [reply Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs:305]: Currently it look like this: Screenshot 2023-11-28 185152
But as written in https://github.com/Flow-Launcher/Flow.Launcher/issues/980#issuecomment-1014272510 it should look like this:
Screenshot 2023-11-28 185040
So the title should only be displayed in the result, but not in the search bar, if this is possible at all :)

Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs:305]: > Currently it look like this: Screenshot 2023-11-28 185152 But as written in [#980 (comment)](https://github.com/Flow-Launcher/Flow.Launcher/issues/980#issuecomment-1014272510) it should look like this: Screenshot 2023-11-28 185040 So the title should only be displayed in the result, but not in the search bar, if this is possible at all :)

that's ok, Its the autocomplete no need to worry

**Comment by VictoriousRaptor** [comment]
@flooxo is this PR ready for further review?

File: Plugins/Flow.Launcher.Plugin.PluginsManager/Languages/en.xaml
Comment (VictoriousRaptor) [inline Plugins/Flow.Launcher.Plugin.PluginsManager/Languages/en.xaml:35]:
```suggestion
    <system:String x:Key="plugin_pluginsmanager_update_all_prompt">Would you like to update {0} plugins?{1}Flow Launcher will restart after updating all plugins.</system:String>
```

File: Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs
Comment (taooceros) [inline Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs:326]: Use the async version instead as this is mainly io

**Review by taooceros — CHANGES_REQUESTED** [review]

File: Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs
Comment (flooxo) [reply Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs:330]: I used the code of a single update as a guide, hence the slight nesting. You are of course absolutely right, it is redundant to start a task in the parallel loop.

File: Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs
Comment (flooxo) [reply Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs:326]: So I should rather use await Task.WhenAll?
because then I get this error and I don't quite know how to solve it.
```
Cannot convert asynchronous lambda expression to delegate type 'Func<ActionContext, bool>'. An asynchronous lambda expression can return void, Task or Task<T>, none of which are convertible to 'Func<ActionContext, bool>'.
```

File: Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs
Comment (taooceros) [inline Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs:329]: Sorry you may understand my idea wrongly. Parallel.ForeachAsync is still probably the best to use.

**Review by taooceros — APPROVED** [review]

**Comment by taooceros** [comment]
@flooxo when testing I find an issue that you didn't collect the LINQ query, which means a multiple enumeration is happening (thanks for rider's hint), and thus result in the final message always say 0 plugins are updated successfully. Otherwise LGTM and thanks for the feature!

**Comment by taooceros** [comment]
Thank you @flooxo !

**Pull request merged**
